### PR TITLE
#3539: Make tool tips indicate the action that a click will acheive

### PR DESCRIPTION
**Pull request by raven2cz**
#3539: Make tool tips indicate the action that a click will acheive

It is implemented according to your comments. 
In addition, there is possibility to define custom tool tips messages in theme and improved visualization of tool tips.

### CODE REVIEW

File: lib/awful/titlebar.lua
Comment (Elv13) [inline lib/awful/titlebar.lua:773]: Can you add 4 `beautiful` variables for these and leave the default as they were before (`nil`). This way it doesn't cause a change in behavior, but can still be defined.

I know there is a lot of precedents to do what you did, but we try not to add more.

File: themes/default/theme.lua
Comment (Elv13) [inline themes/default/theme.lua:101]: If this is a copy of the default one, I would rather not re-define it in the theme. People fork/copy themes rather often. That means if we add new buttons later, they will be stuck with whatever was defined when they copied the theme.

File: lib/awful/titlebar.lua
Comment (raven2cz) [reply lib/awful/titlebar.lua]: Done. Add theme variables and documentation.

File: lib/awful/titlebar.lua
Comment (raven2cz) [reply lib/awful/titlebar.lua]: True. Improved according to your comment.

File: themes/default/theme.lua
Comment (raven2cz) [reply themes/default/theme.lua]: Default theme reverted. Described in documentation with usage example.

**Review by Elv13 — APPROVED** [review]
Thanks for the fixes!

File: lib/awful/titlebar.lua
Comment (raven2cz) [reply lib/awful/titlebar.lua:479]: Yes. This was previous state. Elv13 had same notes, you are in same team :)
So, theme variables are generated by your rules: "titlebar_" .. name .. "_button_" .. prefix .. img .. state. Just check last modifs driven by elv13.

File: lib/awful/titlebar.lua
Comment (raven2cz) [reply lib/awful/titlebar.lua:817]: This is little bit driven by more advanced library "Nice", where is much more better driven placement and timing of tooltip. If you will test the tooltips longer time, you will se that there is problem of clicking to buttons, because tooltip is shown across the button. You have to click to tooltip first, tooltip will be closed, and after that you have chance to click to button. This happens very often. The "nice" plugin solves this problem with better placement of tooltip to the right corner with timing. There is no double click required , you just correct click to button without closing of tooltip. 
It is main reason why I added these configuration parameters, because standard behaviour of default theme is not "correct". So, I tried to open possibility to configure better behaviour which is known in "nice" component. The standard default settings is still kept.

File: lib/awful/titlebar.lua
Comment (raven2cz) [reply lib/awful/titlebar.lua:485]: Yes, this is just link to the documentation of `delay_show` and others similar params which are already described in the `awful.tooltip` documentation part. 
If you want I can copy descriptions from `awful.tooltip` for all config parameters. Or I don't know if there exists some alternative solution to post the link inside the description part? Depends on you. 

In addition, it doesn't belong to this issue, but I saw that all next parameters inside this `titlebar.lua` have some wrong structure for description part, like this one `--- close_button_normal_hover.` This is wrongly shown in the documentation. Can I remove wrong symbols `_` to fix this doc page? What do you think?

File: lib/awful/titlebar.lua
Comment (Aire-One) [reply lib/awful/titlebar.lua:817]: I understand why you made these changes, and I'm happy to see more configuration available here to allow the user to better place the tooltip.

My comment was an open question to understand why you had to change the value for `delay_show` from `1` to `1.0`. Whether there is any technical aspect, I'm not aware of.

File: lib/awful/titlebar.lua
Comment (Aire-One) [reply lib/awful/titlebar.lua:479]: I don't think it's related to @Elv13 comments.

I ask to not use a table, but multiple different string variables. So that instead of `theme.titlebar_tooltip_messages.<variable>_<status>` we would have `theme.titlebar_tooltip_messages_<variable>_<status>`. It would make it more consistent with the existing code base.

File: lib/awful/titlebar.lua
Comment (Aire-One) [reply lib/awful/titlebar.lua:849]: Ok, I get it : `beautiful.titlebar_tooltip_messages` is a table, we have to make sure it exists before trying to access the actual string values.

(Hm, that's honestly something that reinforce my idea that we shouldn't use a table for titlebar_tooltip_messages.)

(If we want to keep this table as is) Could we at least have a util function `get_tooltip_msg(name, img)` to handle these checks and "default waterfall logic"?  I don't see any point where we shouldn't have to do these checks to get the tooltip message, and it would make the code more explicit/self-explaining.

File: lib/awful/titlebar.lua
Comment (Aire-One) [reply lib/awful/titlebar.lua:485]: > If you want I can copy descriptions from awful.tooltip for all config parameters. Or I don't know if there exists some alternative solution to post the link inside the description part? Depends on you

@Elv13, your call. Should we have links to (basically here) function parameters? Like, the `delay_show` property here would link to the `delay_show` parameter inside the `awful.tooltip` constructor. I think it would be technically not so easy since it means we would have to put anchors on every possible function parameters.

I would be ok with "we copy-past the current brief description from the parameter".

> In addition, it doesn't belong to this issue, but I saw that all next parameters inside this titlebar.lua have some wrong structure for description part, like this one --- close_button_normal_hover. This is wrongly shown in the documentation. Can I remove wrong symbols _ to fix this doc page? What do you think?

Yeah, they definitely need some love.

These underscores are necessary here because that's the item name. It fails because the markdown tool from the doc generation tool interpret them as "I want italics here"... I don't know if putting the variable name in backticks would fix that or whether we need to trick them with backslashes... :thinking: 

Maybe better to do that in a separate PR to not flood this feature addition.

(I'm pretty sure this  markdown bug can be linked to https://github.com/awesomeWM/awesome/issues/3500)

File: lib/awful/titlebar.lua
Comment (Aire-One) [inline lib/awful/titlebar.lua:827]: I think this variable declaration here can be moved to the assignation on line 850 since it's not used before.

File: lib/awful/titlebar.lua
Comment (Elv13) [reply lib/awful/titlebar.lua:485]: In the new doc, that would be semi doable. There is already a `<span>` tag in there I could convert to `<a>` without much effort. Plus, it's more organized now, so we could even highlight the row when the anchor is selected. But since that work is far from merged (and still have regressions, so it's not just me being lazy), I don't think this is actionable *now*. This PR can/should get merged before the new doc branch is finished.

image

File: lib/awful/titlebar.lua
Comment (Elv13) [reply lib/awful/titlebar.lua:479]: Sorry if I wasn't clear before. Yes, the names should ideally be flat (no sub-tables). And yes, that's pretty annoying for everybody, but rather hard to fix at this point. I would prefer the names to remain flat as @Aire-One mentions.

The historical reason it is like this is just sad[1], but to break a convention, you have to have a really god reason. Good consistency really helps lower the learning curve.

[1] [Well into the 3.x branch](https://github.com/awesomeWM/awesome/blob/3.1/themes/default/theme.in), the themes were `rc` files rather than `lua` files. This was a leftover from the 1.x/2.x days where Awesome scripting was actually implemented over an RPC in Ruby. To easy the transition (now like 14 years old), the names were kept so the `rc` file could be renamed `lua` with `theme.` added in front of each line,

**Review by Elv13 — CHANGES_REQUESTED** [review]

File: lib/awful/titlebar.lua
Comment (raven2cz) [reply lib/awful/titlebar.lua]: Done. Doc is updated according to this recommendation.

File: lib/awful/titlebar.lua
Comment (raven2cz) [reply lib/awful/titlebar.lua]: Reverted to 1.

File: lib/awful/titlebar.lua
Comment (raven2cz) [reply lib/awful/titlebar.lua]: Doc is improved.

File: lib/awful/titlebar.lua
Comment (raven2cz) [reply lib/awful/titlebar.lua]: Moved to the inner if-condition section.

File: lib/awful/titlebar.lua
Comment (Aire-One) [inline lib/awful/titlebar.lua:887]: nit: Since there is no other usage of this variable, I think it can be moved to the inside of the `if titlebar.enable_tooltip` block.

**Review by Aire-One — APPROVED** [review]
Thank you, @raven2cz, for your work!

(I don't know how we should go about testing for this new feature)

File: lib/awful/titlebar.lua
Comment (raven2cz) [reply lib/awful/titlebar.lua:887]: Maybe I don't understand now the description. Did you mean move whole business logic which is written in the `update()` function to the condition `if titlebar.enable_tooltip then`?

I already tested this before. But it cannot work, because it needs updated parts of `img` and has to be implemented inside `update()` function in the actual place.

**Comment by raven2cz** [comment]
Can you merge this MR? I don't see here any open points. The issue is done.

**Review by Elv13 — DISMISSED** [review]

**Comment by raven2cz** [comment]
@Elv13 So, I see that MR was accepted. Nice. Can you merge it to master? I haven't write access.

**Comment by Aire-One** [comment]
@raven2cz, can you please sync (merge or rebase) your PR branch with master?

We recently merged 2 important PRs:

- The first fixes the Luacheck errors that made the CI fail on the "Build & Test / fixed-lgi-lua5.2 " job.
- The second is the big documentation linting.

So, yeah... Sorry to ask you to do more "useless" work here, but I think it's necessary to sync all PR with master now.

As a nice side effect, we can note that with the Luacheck fixes, the CI now can be green and Mergify should be able to perform automatic merge again.

**Comment by raven2cz** [comment]
> @raven2cz, can you please sync (merge or rebase) your PR branch with master?
> 
> We recently merged 2 important PRs:
> 
>     * The first fixes the Luacheck errors that made the CI fail on the "Build & Test / fixed-lgi-lua5.2 " job.
> 
>     * The second is the big documentation linting.
> 
> 
> So, yeah... Sorry to ask you to do more "useless" work here, but I think it's necessary to sync all PR with master now.
> 
> As a nice side effect, we can note that with the Luacheck fixes, the CI now can be green and Mergify should be able to perform automatic merge again.

Done. But 2 approving reviews are required again. 
I have to trim white spaces and max 120 length for new doc lines.
So, it can be merge now. Sync with master is done. Missing just 2 approves.

**Comment by Aire-One** [comment]
Nice! Thank you for the quick follow up :smile:

**Review by Aire-One — APPROVED** [review]

**Review by Elv13 — APPROVED** [review]

**Comment by Elv13** [comment]
@raven2cz Thanks for the feature

@Aire-One Mergify.io seems broken, it didn't see the PR and didn't auto merge it. This is the first time we had one which should have auto-merged since we moved to GitHub Actions. This will need to be investigated eventually.

**Pull request merged**
